### ISSUE TITLE
Whitespace in <dir>

### ISSUE THREAD

**Issue by rklugseder** [rism-ch/verovio#551]
Rendering problem in MEI-files: dir with whitespace is not rendered correctly.

See example image.
This problem often occurs after a "format document" in Oxygen (ctrl-shft-P). Oxygen "breaks" the lines.

![verovio_dir_whitespace](https://cloud.githubusercontent.com/assets/11180370/25169159/4c80fc24-24e6-11e7-82aa-73f93470b725.jpg)

**Comment by lpugin** [comment]
Is there an option in Oxygen to re-format without breaking the lines?

**Comment by pe-ro** [comment]
If I may jump in here -- oXygen can be told not to break lines within specified elements, but the better option is to normalize the white space (replace multiple spaces with a single one) anytime you want to render textual content of an element.

**Comment by rettinghaus** [comment]
I don't think general normalization of white space would be a good idea. (@pe-ro Perhaps this would be a thing to have an attribute in MEI?)

**Comment by lpugin** [comment]
I looked at this issue at some point and it is actually not straight forward. For example, we cannot just trim text because `<dir>Allegro <sic>molto</sic></dir>` will then be rendered as `Allegromolto`. This is a known issue, see https://wiki.tei-c.org/index.php/XML_Whitespace

Any ideas are welcome.

**Comment by craigsapp** [comment]
Should someone use `<br/>` if they want a literal newline in a `<dir>`?  Or `<lb/>` in TEI world, I suspect :-)

You could replace newlines with spaces...

All spaces are preserved in the SVG output.  Is that desired?  If so, then cleaning newlines would be a problem, since it would add a space where there was not one in the data.  Allowing spaces to be preserved will lead to messy data, since people will want to use it to space the text:

<img width="1055" alt="screen shot 2017-04-19 at 6 38 06 am" src="https://cloud.githubusercontent.com/assets/3487289/25182956/14d0e354-24cb-11e7-9841-829e73bb0e6c.png">

It might be best to add this algorithm to the direction if it is not allowed to be multi-lined:

(1) convert newlines into spaces
(2) convert sequences of multiple spaces into a single space
(3) remove spaces at start and end of dir content.

MEI test data:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<?xml-model href="http://music-encoding.org/schema/3.0.0/mei-all.rng" type="application/xml" schematypens="http://relaxng.org/ns/structure/1.0"?>
<?xml-model href="http://music-encoding.org/schema/3.0.0/mei-all.rng" type="application/xml" schematypens="http://purl.oclc.org/dsdl/schematron"?>
<mei xmlns="http://www.music-encoding.org/ns/mei" meiversion="3.0.0">
    <meiHead>
        <fileDesc>
            <titleStmt>
                <title />
            </titleStmt>
            <pubStmt />
        </fileDesc>
        <encodingDesc>
            <appInfo>
                <application isodate="2017-04-19T06:38:04" version="0.9.14-dev-d2d89fc-dirty">
                    <name>Verovio</name>
                    <p>Transcoded from Humdrum</p>
                </application>
            </appInfo>
        </encodingDesc>
        <workDesc>
            <work>
                <titleStmt>
                    <title />
                </titleStmt>
            </work>
        </workDesc>
    </meiHead>
    <music>
        <body>
            <mdiv>
                <score>
                    <scoreDef xml:id="scoredef-0000000767210337">
                        <staffGrp xml:id="staffgrp-0000002040714675">
                            <staffDef xml:id="staffdef-0000002122344513" n="1" clef.shape="G" clef.line="2" meter.count="4" meter.unit="4" lines="5" />
                        </staffGrp>
                    </scoreDef>
                    <section xml:id="section-0000000372031839">
                        <measure xml:id="measure-L3" n="1" type="m-1">
                            <staff xml:id="staff-L3F1N1" n="1">
                                <layer xml:id="layer-L3F1N1" n="1">
                                    <note xml:id="note-L5F1" type="qon-0 qoff-4 b40-162" dur="1" oct="4" pname="c">
                                        <accid xml:id="accid-L5F1" accid.ges="n" />
                                    </note>
                                </layer>
                            </staff>
                            <dir xml:id="dir-L5F1" place="above" staff="1" tstamp="1.000000">a b c</dir>
                        </measure>
                        <measure xml:id="measure-L6" n="2" type="m-2">
                            <staff xml:id="staff-L6F1N1" n="1">
                                <layer xml:id="layer-L6F1N1" n="1">
                                    <note xml:id="note-L8F1" type="qon-4 qoff-8 b40-168" dur="1" oct="4" pname="d">
                                        <accid xml:id="accid-L8F1" accid.ges="n" />
                                    </note>
                                </layer>
                            </staff>
                            <dir xml:id="dir-L8F1" place="above" staff="1" tstamp="1.000000">a  b  c</dir>
                        </measure>
                        <measure xml:id="measure-L9" n="3" type="m-3">
                            <staff xml:id="staff-L9F1N1" n="1">
                                <layer xml:id="layer-L9F1N1" n="1">
                                    <note xml:id="note-L11F1" type="qon-8 qoff-12 b40-174" dur="1" oct="4" pname="e">
                                        <accid xml:id="accid-L11F1" accid.ges="n" />
                                    </note>
                                </layer>
                            </staff>
                            <dir xml:id="dir-L11F1" place="above" staff="1" tstamp="1.000000">a   b   c</dir>
                        </measure>
                        <measure xml:id="measure-L12" n="4" type="m-4">
                            <staff xml:id="staff-L12F1N1" n="1">
                                <layer xml:id="layer-L12F1N1" n="1">
                                    <note xml:id="note-L14F1" type="qon-12 qoff-16 b40-179" dur="1" oct="4" pname="f">
                                        <accid xml:id="accid-L14F1" accid.ges="n" />
                                    </note>
                                </layer>
                            </staff>
                            <dir xml:id="dir-L14F1" place="above" staff="1" tstamp="1.000000">a    b    c</dir>
                        </measure>
                        <measure xml:id="measure-L15" n="5" right="end" type="m-5">
                            <staff xml:id="staff-L15F1N1" n="1">
                                <layer xml:id="layer-L15F1N1" n="1">
                                    <note xml:id="note-L17F1" type="qon-16 qoff-20 b40-185" dur="1" oct="4" pname="g">
                                        <accid xml:id="accid-L17F1" accid.ges="n" />
                                    </note>
                                </layer>
                            </staff>
                            <dir xml:id="dir-L17F1" place="above" staff="1" tstamp="1.000000">a     b     c</dir>
                        </measure>
                    </section>
                </score>
            </mdiv>
        </body>
    </music>
</mei>
```

**Comment by pe-ro** [comment]
@lpugin: The TEI page you referenced says what I had in mind -- normalization = collapse + trim.  I think MEI encoders should assume this will be the default.  But in no case should `<dir>Allegro <sic>molto</sic></dir>` ever be rendered as `Allegromolto`.

@rettinghaus: MEI already allows has a `att.whitespace` attribute class which includes `@xml:space`.  Currently, its only members are: `<extMeta>`, `<incipCode>`, and `<rend>`.  It could be added to other elements, but I'd rather not.  I'd prefer to encourage the use of `<rend>`, e.g., 
```
<dir tstamp="1">
  <rend xml:space="preserve">Allegro       <sic>molto  </sic>  </rend>
</dir>
```
Since `<rend>` allows all text phrase-level elements, this shouldn't present any problems. It will preserve the space following "Allegro" and that following "molto", but not the new lines inside `<dir>`.  

@craigsapp: Encoders may preserve space if they want to.  Notice I didn't say it's a good idea. 😀

**Comment by lpugin** [comment]
> But in no case should `<dir>Allegro <sic>molto</sic></dir>` ever be rendered as `Allegromolto`.

Yes, this is the most tricky part. 
https://wiki.tei-c.org/index.php/XML_Whitespace#Mixed-Content_Elements\

It is currently handled here.
https://github.com/rism-ch/verovio/blob/master/src/iomei.cpp#L2868-L2872

Additional normalisation should be done in `MeiInput::ReadMeiText`

**Comment by pe-ro** [comment]
I suppose mixed content can be tricky, but it's a feature of XML, not a defect. 😀

**Comment by lpugin** [comment]
Fixed with #552
